File: c++/68_Text_Justification.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
        vector<string> res;
        vector<string> stk;
        string line = "";
        if (maxWidth == 0){
            res.push_back(line);
            return res;
        }
        int totalLen = 0;
        for (int i = 0; i < words.size(); i++) {
            int wlen = words[i].size();
            if (i > 0 and totalLen + wlen + stk.size() > maxWidth){
                if (stk.size() > 1) {
                    int lenSpaces = maxWidth - totalLen;
                    if (stk.size() > 2) {
                        int lenavg = lenSpaces / (stk.size() - 1);
                        int lenArr[stk.size() - 1] = {0};
                        int firstNum = lenSpaces - lenavg * (stk.size() - 1);
                        string tmp(lenavg, ' ');
                        for (int j = 0; j < firstNum; j++)
                            line += stk[j] + tmp + ' ';
                        for (int j = firstNum; j < stk.size() - 1; j++)
                            line += stk[j] + tmp;
                        line += stk.back();
                    }
                    else {
                        string tmp(lenSpaces, ' ');
                        line = stk[0] + tmp + stk[1];
                    }
                }
                else{
                    string tmp((maxWidth - stk[0].size()), ' ');
                    line = stk[0] + tmp;
                }
                res.push_back(line);
                line = "";
                totalLen = 0;
                stk.clear();
            }
            totalLen += wlen;
            stk.push_back(words[i]);
        }
        if (stk.size() == 1) {
            string tmp((maxWidth - stk[0].size()), ' ');
            line = stk[0] + tmp;
        }
        else{
            for(int i = 0; i < stk.size(); i++)
                line += stk[i] + " ";
            string tmp((maxWidth - line.size()), ' ');
            line += tmp;
        }
        res.push_back(line);
        return res;
    }
};


int main() {
	Solution s;
    // ["What","must","be","shall","be."]
    string a1 = "what";
    string a2 = "must";
    string a3 = "be";
    string a4 = "shall";
    string a5 = "be.";
    vector<string> words;
    words.push_back(a1);
    words.push_back(a2);
    words.push_back(a3);
    words.push_back(a4);
    words.push_back(a5);
    vector<string> res = s.fullJustify(words, 12);

    cout << res[0] << endl;

}

File: python/68_Text_Justification.py
#!/usr/bin/python
import unittest



class Solution(object):
    def fullJustify(self, words, maxWidth):
        if maxWidth == 0: return [""]
        res = []
        stk = []
        line = ""
        totalLen = 0
        for i, word in enumerate(words):
            wlen = len(word)
            if i > 0 and totalLen + wlen + len(stk) > maxWidth:
                if len(stk) > 1:
                    lenSpaces = maxWidth - totalLen # number of spaces left
                    if len(stk) > 2:
                        lenavg = lenSpaces / (len(stk) - 1)
                        lenArr = [0] * (len(stk) - 1)
                        firstNum = lenSpaces - lenavg * (len(stk) - 1)
                        for i in range(firstNum):
                            line += stk[i] + " " * (lenavg + 1)
                        for i in range(firstNum, len(stk) - 1):
                            line += stk[i] + " " * lenavg
                        line += stk[-1]
                    else:
                        line = stk[0] + " " * lenSpaces + stk[1]
                else:
                    line = stk[0] + " " * (maxWidth - len(stk[0]))
                res.append(line)
                line = ""
                totalLen = 0
                stk = []
            totalLen += wlen
            stk.append(word)
        # deal with the last word
        if len(stk) == 1:
            line = stk[0] + " " * (maxWidth - len(stk[0]))
        else:
            for key in stk:
                line += key + " "
            line += " " * (maxWidth - len(line))
        res.append(line)
        return res

def main():
    s = Solution()
    #words = ["This", "is", "an", "example", "of", "text", "justification."]
    words = ["Don't","go","around","saying","the","world","owes","you","a","living;","the","world","owes","you","nothing;","it","was","here","first."]
    maxWidth = 30
    print s.fullJustify(words, maxWidth)


if __name__ == "__main__":
    main()
